homeassistant/components/switch/rpi_servo.py

```diff
@@ -0,0 +1,127 @@
+                 default=DEFAULT_POSITION_ON): cv.positive_int,
+    vol.Optional(CONF_POSITION_OFF,
+                 default=DEFAULT_POSITION_OFF): cv.positive_int,
+    vol.Optional(CONF_POSITION_DURATION,
+                 default=DEFAULT_POSITION_DURATION): cv.positive_int,
+})
+
+
+# pylint: disable=unused-argument
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Raspberry PI GPIO servo devices."""
+    inactive_position = config.get(CONF_INACTIVE_POSTION)
+    position_on = config.get(CONF_POSITION_ON)
+    position_off = config.get(CONF_POSITION_OFF)
+    enabled_duration = config.get(CONF_POSITION_DURATION)
+    switches = []
+    ports = config.get(CONF_PORTS)
+    for port, name in ports.items():
+        switches.append(RPiGPIOServo(hass, name, port, inactive_position,
+                                     position_on, position_off,
+                                     enabled_duration))
+    add_devices(switches)
+
+
+class RPiGPIOServo(ToggleEntity):
+    """Representation of a  Raspberry Pi GPIO Servo."""
+
+    def __init__(self, hass, name, port, inactive_position, position_on,
+                 position_off, enabled_duration):
+        """Initialize the servo."""
+        self._hass = hass
+        self._name = name or DEVICE_DEFAULT_NAME
+        self._port = port
+        self._state = False
+        self._inactive_position = inactive_position
+        self._position_on = position_on
+        self._position_off = position_off
+        self._enabled_duration = enabled_duration
+        gpio.setup_output(port)
+
+    @property
+    def name(self):
+        """Return the name of the switch."""
+        return self._name
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._state
+
+    def turn_on(self, **kwargs):
+        """Turn the switch on."""
+        duty = get_duty(self._position_on)
+        self._run_with_duty(duty)
+        self._state = True
+        self.schedule_update_ha_state()
+
+    def turn_off(self, **kwargs):
+        """Turn the switch off."""
+        duty = get_duty(self._position_off)
+        self._run_with_duty(duty)
+        self._state = False
+        self.schedule_update_ha_state()
+
+    def _run_with_duty(self, duty):
+        """Turn the servo to the given duty."""
+        if self._enabled_duration == 0:
+            gpio.run_pwm(self._hass, self._port, 50, duty, 2)
+        else:
+            gpio.run_pwm(self._hass, self._port, 50, duty,
+                         self._enabled_duration)
+            duty = get_duty(self._inactive_position)
+
+            def inactivate(*args):
+                """Return the servo to inactive position."""
+                gpio.run_pwm(self._port, 50, duty, self._enabled_duration)
+
+            track_point_in_time(self._hass, inactivate, self._enabled_duration)
```

Comment:
See my comment above about callback. If you implement my suggestion, you would remove this scheduling.